agent/engine/containermetadata/write_metadata.go

```diff
@@ -0,0 +1,267 @@
+func removeContents(dir string) error {
+	d, err := os.Open(dir)
+	if err != nil {
+		return err
+	}
+	defer d.Close()
+	names, err := d.Readdirnames(-1)
+	if err != nil {
+		return nil
+	}
+	for _, name := range names {
+		err = os.RemoveAll(filepath.Join(dir, name))
+		if err != nil {
+			return err
+		}
+	}
+	return os.Remove(dir)
+}
+
+// MetadataManager is an interface that allows us to abstract away the metadata
+// operations
+type MetadataManager interface {
+	CreateMetadata([]string, *api.Task, *api.Container) ([]string, error)
+	UpdateMetadata(string, *api.Task, *api.Container) error
+	CleanTaskMetadata(*api.Task) error
+}
+
+// metadataManager implements the MetadataManager interface
+type metadataManager struct {
+	client dockerDummyClient
+	cfg    *config.Config
+}
+
+// NewMetadataManager creates a metadataManager for a given DockerTaskEngine settings.
+func NewMetadataManager(client dockerDummyClient, cfg *config.Config) MetadataManager {
+	manager := &metadataManager{
+		client: client,
+		cfg:    cfg,
+	}
+	return manager
+}
+
+// CreateMetadata creates the metadata file and adds the metadata directory to
+// the container's mounted host volumes
+// binds []string is passed by value to avoid race conditions by multiple
+// calls to CreateMetadata, although this should never actually happen
+func (manager *metadataManager) CreateMetadata(binds []string, task *api.Task, container *api.Container) ([]string, error) {
+	// Check if manager has invalid entries
+	var err error
+	if manager.cfg == nil {
+		err = fmt.Errorf("Failed to create metadata: Invalid inputs")
+		return binds, err
+	}
+
+	// Do not create metadata file for internal containers
+	// Add error handling for this case? Probably no need since
+	// Internal containers should not be visible to users anyways
+	if container.IsInternal {
+		return binds, nil
+	}
+
+	// Create task and container directories if they do not yet exist
+	mdDirectoryPath := getMetadataFilePath(task, container, manager.cfg.DataDir)
+	// Stop metadata creation if path is malformed for any reason
+	if mdDirectoryPath == "" {
+		err = fmt.Errorf("Failed to create metadata: Invalid file path")
+		return binds, err
+	}
+
+	err = os.MkdirAll(mdDirectoryPath, os.ModePerm)
+	if err != nil {
+		err = fmt.Errorf("Failed to create metadata directory at %s: %s", mdDirectoryPath, err.Error())
+		return binds, err
+	}
+
+	// Create metadata file
+	mdFilePath := filepath.Join(mdDirectoryPath, metadataFile)
+	err = ioutil.WriteFile(mdFilePath, nil, 0644)
+	if err != nil {
+		err = fmt.Errorf("Failed to create metadata file at %s: %s", mdFilePath, err.Error())
+		return binds, err
+	}
+
+	// Acquire the metadata then write it in JSON format to the file
+	metadata := acquireMetadataAtContainerCreate(manager.client, manager.cfg, task)
+	err = metadata.writeToMetadataFile(task, container, manager.cfg.DataDir)
+	if err != nil {
+		err = fmt.Errorf("Failed to write to metadata file %s: %s", mdFilePath, err.Error())
```

Comment:
I was trying to imitate the errors.Wrap format. Should there be a colon after "error"?